File: typescript/src/sortArray/sortArray.ts
export const sortArray = (nums: number[]): number[] => {
    return mergeSort(nums);
};

const mergeSort = (nums: number[]): number[] => {
    if(nums.length == 1){
        return nums;
    }
    const mid = Math.floor(nums.length/2);
    const leftArray = nums.slice(0, mid);
    const rightArray = nums.slice(mid, nums.length);

    const left = mergeSort(leftArray);
    const right = mergeSort(rightArray);
    return merge(left, right);
}

const merge = (leftArray: number[], rightArray: number[]): number[] => {
    const length = leftArray.length + rightArray.length;
    let left = 0;
    let right = 0;
    const nums: number[] = []; 
    let i = 0;

    while(left < leftArray.length && right < rightArray.length){
        if(leftArray[left] < rightArray[right]){
            nums[i] = leftArray[left];
            left++;
        } else {
            nums[i] = rightArray[right];
            right++;
        }
        i++;
    }

    while(left < leftArray.length){
        nums[i] = leftArray[left];
        left++;
        i++;
    }

    while(right < rightArray.length){
        nums[i] = rightArray[right];
        right++;
        i++;
    }

    return nums;
}

const nums = [2,7,11,4,5,3,9,17,14];
console.log(sortArray(nums));

File: go/sortArray/sortArray.go
package main

import (
	"math"
)

func SortArray(nums []int) []int {
	return mergeSort(nums)
}

func mergeSort(nums []int) []int {
	length := len(nums)
	if length == 1 {
		return nums
	}
	mid := int(math.Floor(float64(length / 2)))
	leftArray := nums[0:mid]
	rightArray := nums[mid:length]

	left := mergeSort(leftArray)
	right := mergeSort(rightArray)

	return merge(left, right)
}

func merge(leftArray []int, rightArray []int) []int {
	left := 0
	right := 0
	i := 0
	nums := make([]int, len(leftArray)+len(rightArray))

	for left < len(leftArray) && right < len(rightArray) {
		if leftArray[left] < rightArray[right] {
			nums[i] = leftArray[left]
			left += 1
		} else {
			nums[i] = rightArray[right]
			right += 1
		}
		i += 1
	}

	for left < len(leftArray) {
		nums[i] = leftArray[left]
		left += 1
		i += 1
	}
	for right < len(rightArray) {
		nums[i] = rightArray[right]
		right += 1
		i += 1
	}

	return nums
}
